ts/src/base/Exchange.ts

```diff
@@ -4516,6 +4516,147 @@ export default class Exchange {
+                    if (responseLength < maxEntriesPerRequest) {
+                        break;
+                    }
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    lastTimestamp = this.safeInteger (responseLength - 1, 'timestamp');
+                }
+            } catch (e) {
+                if (errors + 1 > maxRetries) {
+                    throw e;
+                }
+                errors = errors + 1;
+            }
+        }
+        return this.removeRepeatedElementsFromArray (result);
+    }
+
+    async fetchPaginatedCallDeterministic (method: string, symbol: string = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}, maxEntriesPerRequest = 1000): Promise<any> {
+        let maxCalls = undefined;
+        [ maxCalls, params ] = this.handleOptionAndParams (params, method, 'paginationCalls', 20);
+        const now = this.milliseconds ();
+        const tasks = [];
+        const time = this.parseTimeframe (timeframe) * 1000;
+        const step = time * maxEntriesPerRequest;
+        let currentSince = now - (maxCalls * step) - 1;
+        if (since !== undefined) {
+            currentSince = Math.max (currentSince, since);
+        }
+        for (let i = 0; i < maxCalls; i++) {
+            if (timeframe && method !== 'fetchFundingRateHistory') {
+                tasks.push (this[method] (symbol, timeframe, currentSince, maxEntriesPerRequest, params));
+            } else {
+                tasks.push (this[method] (symbol, currentSince, maxEntriesPerRequest, params));
+            }
+            currentSince = currentSince + step - 1;
+        }
+        const results = await Promise.all (tasks);
+        let result = [];
+        for (let i = 0; i < results.length; i++) {
+            result = result.concat (results[i]);
+        }
+        return this.removeRepeatedElementsFromArray (result) as any;
+    }
+
+    async fetchPaginatedCallCursor (method: string, symbol: string = undefined, since = undefined, limit = undefined, params = {}, cursorKey = undefined, maxEntriesPerRequest = 1000): Promise<any> {
+        let maxCalls = undefined;
+        [ maxCalls, params ] = this.handleOptionAndParams (params, method, 'paginationCalls', 20);
+        let maxRetries = undefined;
+        [ maxRetries, params ] = this.handleOptionAndParams (params, method, 'maxRetries', 3);
+        let cursorValue = undefined;
+        let i = 0;
+        let errors = 0;
+        let result = [];
+        while (i < maxCalls) {
+            try {
+                if (cursorValue !== undefined) {
+                    params[cursorKey] = cursorValue;
+                }
+                const response = await this[method] (symbol, since, maxEntriesPerRequest, params);
+                errors = 0;
+                if (this.verbose) {
+                    console.log ('Pagination call', i, 'response length', response.length);
```

Comment:
should the log be also included in the fetchPaginatedCallDynamic?

Reply:
good idea, that would be useful yes